File: typescript/src/webhook.ts
import { createHmac, timingSafeEqual } from "node:crypto";
import type { VerifyWebhookSignatureParams } from "./types.js";

/**
 * Verifies the authenticity and freshness of an Escalation Engine webhook event.
 *
 * The signature header has the format: `t=<unix-timestamp>,v1=<hex-hmac>`
 *
 * The signed payload is constructed as: `<timestamp>.<raw-body>`
 *
 * This function:
 * 1. Parses the timestamp and HMAC from the signature header.
 * 2. Recomputes the HMAC-SHA256 using the webhook secret.
 * 3. Compares the signatures using a timing-safe comparison.
 * 4. Rejects signatures older than the tolerance window (default: 300s).
 *
 * @param params - Verification parameters.
 * @returns `true` if the signature is valid and fresh, `false` otherwise.
 *
 * @example
 * ```typescript
 * import { verifyWebhookSignature } from '@escalation-engine/sdk';
 *
 * const isValid = verifyWebhookSignature({
 *   payload: req.body,       // raw request body string
 *   signature: req.headers['x-escalation-signature'],
 *   secret: process.env.ESCALATION_WEBHOOK_SECRET,
 *   tolerance: 300,
 * });
 * ```
 */
export function verifyWebhookSignature(
  params: VerifyWebhookSignatureParams,
): boolean {
  const { payload, signature, secret, tolerance = 300 } = params;

  if (!payload || !signature || !secret) {
    return false;
  }

  // Parse the signature header: t=<timestamp>,v1=<hmac>
  const parts = signature.split(",");
  const timestampPart = parts.find((p) => p.startsWith("t="));
  const signaturePart = parts.find((p) => p.startsWith("v1="));

  if (!timestampPart || !signaturePart) {
    return false;
  }

  const timestamp = timestampPart.slice(2); // Remove "t="
  const expectedSignature = signaturePart.slice(3); // Remove "v1="

  if (!timestamp || !expectedSignature) {
    return false;
  }

  const timestampNum = Number.parseInt(timestamp, 10);
  if (Number.isNaN(timestampNum)) {
    return false;
  }

  // Check timestamp tolerance to prevent replay attacks
  const now = Math.floor(Date.now() / 1000);
  const age = Math.abs(now - timestampNum);
  if (age > tolerance) {
    return false;
  }

  // Compute expected HMAC: HMAC-SHA256(secret, "<timestamp>.<payload>")
  const signedPayload = `${timestamp}.${payload}`;
  const computedHmac = createHmac("sha256", secret)
    .update(signedPayload)
    .digest("hex");

  // Timing-safe comparison to prevent timing attacks
  try {
    const a = Buffer.from(computedHmac, "hex");
    const b = Buffer.from(expectedSignature, "hex");

    if (a.length !== b.length) {
      return false;
    }

    return timingSafeEqual(a, b);
  } catch {
    return false;
  }
}

/**
 * Constructs a webhook signature header value for testing purposes.
 * Do NOT use this in production — it is provided for writing tests.
 *
 * @param payload - The raw request body string.
 * @param secret - The webhook signing secret.
 * @param timestamp - Optional unix timestamp (defaults to now).
 * @returns The signature string in the format `t=<timestamp>,v1=<hmac>`.
 */
export function constructWebhookSignature(
  payload: string,
  secret: string,
  timestamp?: number,
): string {
  const ts = timestamp ?? Math.floor(Date.now() / 1000);
  const signedPayload = `${ts}.${payload}`;
  const hmac = createHmac("sha256", secret)
    .update(signedPayload)
    .digest("hex");
  return `t=${ts},v1=${hmac}`;
}

File: python/humanrail/webhook.py
"""
Webhook signature verification for the HumanRail.

The HumanRail signs webhook payloads using HMAC-SHA256 with a
per-organization webhook secret. The signature header includes a timestamp
to prevent replay attacks.

Signature header format: `t=<unix-timestamp>,v1=<hex-encoded HMAC>`
Signed payload format: `<timestamp>.<raw-body>`
"""

from __future__ import annotations

import hashlib
import hmac
import time


def verify_webhook_signature(
    *,
    payload: str,
    signature: str,
    secret: str,
    tolerance: int = 300,
) -> bool:
    """Verify the authenticity and freshness of an HumanRail webhook event.

    This function:
    1. Parses the timestamp and HMAC from the signature header.
    2. Recomputes the HMAC-SHA256 using the webhook secret.
    3. Compares the signatures using a timing-safe comparison.
    4. Rejects signatures older than the tolerance window.

    Args:
        payload: The raw request body as a string. Must be the exact bytes
            received, not a re-serialized JSON object.
        signature: The value of the `x-escalation-signature` header.
            Format: ``t=<timestamp>,v1=<hex-encoded HMAC>``
        secret: The webhook signing secret for your organization.
        tolerance: Maximum age of the signature in seconds. Signatures older
            than this are rejected to prevent replay attacks. Defaults to 300 (5 minutes).

    Returns:
        True if the signature is valid and fresh, False otherwise.

    Example::

        from humanrail import verify_webhook_signature

        is_valid = verify_webhook_signature(
            payload=request.body.decode("utf-8"),
            signature=request.headers["x-escalation-signature"],
            secret=os.environ["ESCALATION_WEBHOOK_SECRET"],
            tolerance=300,
        )
    """
    if not payload or not signature or not secret:
        return False

    # Parse the signature header: t=<timestamp>,v1=<hmac>
    parts = signature.split(",")
    timestamp_part: str | None = None
    signature_part: str | None = None

    for part in parts:
        if part.startswith("t="):
            timestamp_part = part[2:]
        elif part.startswith("v1="):
            signature_part = part[3:]

    if not timestamp_part or not signature_part:
        return False

    try:
        timestamp_num = int(timestamp_part)
    except ValueError:
        return False

    # Check timestamp tolerance to prevent replay attacks
    now = int(time.time())
    age = abs(now - timestamp_num)
    if age > tolerance:
        return False

    # Compute expected HMAC: HMAC-SHA256(secret, "<timestamp>.<payload>")
    signed_payload = f"{timestamp_part}.{payload}"
    computed_hmac = hmac.new(
        secret.encode("utf-8"),
        signed_payload.encode("utf-8"),
        hashlib.sha256,
    ).hexdigest()

    # Timing-safe comparison to prevent timing attacks
    return hmac.compare_digest(computed_hmac, signature_part)


def construct_webhook_signature(
    payload: str,
    secret: str,
    timestamp: int | None = None,
) -> str:
    """Construct a webhook signature header value for testing purposes.

    Do NOT use this in production. It is provided for writing tests.

    Args:
        payload: The raw request body string.
        secret: The webhook signing secret.
        timestamp: Optional unix timestamp (defaults to now).

    Returns:
        The signature string in the format ``t=<timestamp>,v1=<hmac>``.
    """
    ts = timestamp if timestamp is not None else int(time.time())
    signed_payload = f"{ts}.{payload}"
    hmac_hex = hmac.new(
        secret.encode("utf-8"),
        signed_payload.encode("utf-8"),
        hashlib.sha256,
    ).hexdigest()
    return f"t={ts},v1={hmac_hex}"
